### PR TITLE
Drop dependency on k8s.io/kubernetes

### PR DESCRIPTION
**Pull request by howardjohn**
https://github.com/helm/helm/issues/6606

Depending on k8s.io/kubernetes is not recommended by Kubernetes, and
forces dependencies of Helm to also depend on them. We are only using
this dependency in one relatively isolated occurance, which can be
easily copied over rather than depending on the entire Kubernetes.

Copying this code is not very desirable, so if we don't want to have
this duplication we can at least use this PR as a PoC and see if we can
get Kubernetes to publish the controller package as a separate Go module
(see
https://github.com/kubernetes/kubernetes/issues/79384#issuecomment-538740756)

Closes https://github.com/helm/helm/issues/6606

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by bacongobbler** [comment]
This doesn't seem legally sound... Unless I'm mistaken, deploymentutil.go is copied from Kubernetes code, so we cannot claim copyright ownership of the file. This seems very similar to the same situation described in #5038.

Similar to https://github.com/helm/helm/pull/5038#issuecomment-474133510, can you please

1. move `/pkg/kube/deploymentutil.go` to a new directory: `/third_party/k8s.io/kubernetes/deployment/util/deployment_util.go` or something indicative that the code came from Kubernetes
2. Please place the full Apache license header comments from the source file. The end result should look identical to [the code in Kubernetes](https://github.com/kubernetes/kubernetes/blob/e856613dd5bb00bcfaca6974431151b5c06cbed5/pkg/controller/deployment/util/deployment_util.go)
3. Add a comment stating where [the original source code exists](https://github.com/kubernetes/kubernetes/blob/e856613dd5bb00bcfaca6974431151b5c06cbed5/pkg/controller/deployment/util/deployment_util.go)
4. If there are changes made to the original deployment_util.go file, please also add comments stating what changed, and why.

This way we have all our legal bases covered as advised from the CNCF legal team.

Thanks!

File: pkg/chartutil/expand.go [hunk 1]
```diff
@@ -45,9 +45,6 @@ func Expand(dir string, r io.Reader) error {
 			if err := yaml.Unmarshal(file.Data, ch); err != nil {
 				return errors.Wrap(err, "cannot load Chart.yaml")
 			}
-			if err != nil {
```
Comment (bacongobbler) [inline pkg/chartutil/expand.go]: would you mind reverting this? Seems like this isn't relevant to the change at hand.

File: pkg/kube/deploymentutil.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
+/*
+Copyright The Helm Authors.
```
Comment (bacongobbler) [inline pkg/kube/deploymentutil.go]: Here's the legal issue I'm referring to.

**Review by bacongobbler — CHANGES_REQUESTED** [review]

**Comment by howardjohn** [comment]
> This doesn't seem legally sound... Unless I'm mistaken, deploymentutil.go is copied from Kubernetes code, so we cannot claim copyright ownership of the file. This seems very similar to the same situation described in #5038.

I was wondering about this. Thanks for the guidance on how to make this work, I'll send a fixup soon

File: pkg/chartutil/expand.go (hunk 1 above)
Comment (howardjohn) [reply pkg/chartutil/expand.go]: Hmm looks like it was a lint error, I must have a different version locally. I'll revert

**Review by jlegrone — COMMENTED** [review]
We should probably avoid making this part of Helm's public API surface. Can we move `pkg/kube` to `internal/kube`? Ref: https://github.com/golang-standards/project-layout/tree/master/internal

Thanks for taking the initiative on this, I think this change will render some hacks I've had to do in the past obsolete!

**Comment by bacongobbler** [comment]
`pkg/kube` has received several enhancements that make it useful to others trying to create/update/delete/watch objects similar to Helm, especially after #6085. Is there a particular reason it should it be made internal?

**Comment by jlegrone** [comment]
Oh right, I forgot that `pkg/kube` already existed. I do think we should consider keeping the methods in `pkg/kube/deploymentutil.go` private, or move the file to an internal package.

The rationale is that we can choose to make these new functions public at any time, but while they're internal we can always refactor to depend on `kubernetes/kubernetes` if this functionality becomes consumable in the future, without introducing a breaking change.

**Comment by bacongobbler** [comment]
Good point.

@howardjohn, instead of `/third_party/k8s.io/kubernetes/deployment/util/deployment_util.go`, can you please move it to `/internal/third_party/k8s.io/kubernetes/deployment/util/deployment_util.go`? Much appreciated.

**Comment by howardjohn** [comment]
@bacongobbler I replaced the Helm license with the Kubernetes license and the license check fails. I added an exception to filter out `third_party`. Let me know if I should do something else instead, like adding both licenses

**Review by bacongobbler — APPROVED** [review]
LGTM! Thanks for doing this. Holding off on merging until another maintainer approves this as well, given the size and legal issues surrounding this.

**Review by jlegrone — APPROVED** [review]

**Comment by howardjohn** [comment]
Is there anything else needed to get this merged?

**Comment by bacongobbler** [comment]
@jlegrone is not a maintainer. Needs another LGTM from a project maintainer of Helm. I'll ping someone today to take a look at this.

**Comment by howardjohn** [comment]
Sounds good, thanks!

**Review by hickeyma — APPROVED** [review]
I am not happy with the way this has to be done because it is quite dirty to have to copy over a file in this way. However, if it is the only way to work around it in the circumstance then it is what it is.

Thanks for the work on this @howardjohn. Good tracking down.

**Comment by sslavic** [comment]
Please consider applying equivalent change to `release-2.15` too. That would make migration easier, decouple steps - adopting Go modules and switch to helm v3 once available.

**Comment by bacongobbler** [comment]
it won't be possible without breaking backwards compatibility, unfortunately. Many contributors have tried, which is why we had to target this for Helm 3.

**Comment by sslavic** [comment]
I guess answer is same for release-2.16, correct?
Curious about backward compatibility being broken, this change internalizes a testing support package and removes no longer used transitive dependency on k8s.io/kubernets; if someone using helm as library used also directly k8s.io/kubernetes without specifying it as direct dependency, depending on helm to bring it in transitively, they should be able to easily fix their build tool dependency configuration and declare direct dependency to k8s.io/kubernetes - what am I missing here?

**Comment by bacongobbler** [comment]
You're right that this change only makes one change to the public API for Helm 3. In Helm 2, the changes are a little [more involved](https://github.com/helm/helm/blob/0ca95341c19fcaf9cecdf50caef1693568835a84/pkg/kube/client.go#L60-L61), which makes it incredibly difficult to remove that dependency without introducing breaking changes.

Hope this helps!

**Pull request merged**